javalib/src/main/scala/java/lang/Math.scala

```diff
@@ -1,324 +1,347 @@
+    if (overflow.flag) throw new ArithmeticException("Long overflow")
+    else overflow.value
   }
 
-  def addExact(a: scala.Int, b: scala.Int): scala.Int = {
-    val res       = a + b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b < 0)) res
-    else throw new ArithmeticException("Integer overflow")
+  @inline def negateExact(a: scala.Int): scala.Int =
+    subtractExact(0, a)
+
+  @inline def negateExact(a: scala.Long): scala.Long =
+    subtractExact(0, a)
+
+  def nextAfter(a: scala.Float, b: scala.Double): scala.Float = {
+    if (Float.isNaN(a) || Double.isNaN(b)) {
+      Float.NaN
+    } else if (abs(a) == 0f && abs(b) == 0d) {
+      b.toFloat
+    } else if (abs(a) == Float.MIN_VALUE && abs(b) < abs(a.toDouble)) {
+      copySign(0, a)
+    } else if (Float.isInfinite(a) && abs(b) < abs(a.toDouble)) {
+      copySign(Float.MAX_VALUE, a)
+    } else if (abs(a) == Float.MAX_VALUE && abs(b) > abs(a.toDouble)) {
+      copySign(Float.POSITIVE_INFINITY, a)
+    } else {
+      cmath.nextafter(a.toDouble, b).toFloat
+    }
   }
 
-  def addExact(a: scala.Long, b: scala.Long): scala.Long = {
-    val res       = a + b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b < 0)) res
-    else throw new ArithmeticException("Long overflow")
+  def nextAfter(a: scala.Double, b: scala.Double): scala.Double = {
+    if (Double.isNaN(a) || Double.isNaN(b)) {
+      Double.NaN
+    } else if (abs(a) == 0f && abs(b) == 0d) {
+      b
+    } else if (abs(a) == Double.MIN_VALUE && abs(b) < abs(a)) {
+      copySign(0, a)
+    } else if (Double.isInfinite(a) && abs(b) < abs(a)) {
+      copySign(Double.MAX_VALUE, a)
+    } else if (abs(a) == Double.MAX_VALUE && abs(b) > abs(a)) {
+      copySign(Double.POSITIVE_INFINITY, a)
+    } else {
+      cmath.nextafter(a, b)
+    }
   }
 
-  def subtractExact(a: scala.Int, b: scala.Int): scala.Int = {
-    val res       = a - b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b > 0)) res
-    else throw new ArithmeticException("Integer overflow")
+  def nextDown(a: scala.Float): scala.Float = {
+    if (Float.isNaN(a)) {
+      Float.NaN
+    } else if (a == Float.NEGATIVE_INFINITY) {
+      Float.NEGATIVE_INFINITY
+    } else if (abs(a) == 0f) {
+      -Float.MIN_VALUE
+    } else {
+      cmath.nextafterf(a, Float.NEGATIVE_INFINITY)
+    }
   }
 
-  def subtractExact(a: scala.Long, b: scala.Long): scala.Long = {
-    val res       = a - b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b > 0)) res
-    else throw new ArithmeticException("Long overflow")
+  def nextDown(a: scala.Double): scala.Double = {
+    if (Double.isNaN(a)) {
+      Double.NaN
+    } else if (a == Double.NEGATIVE_INFINITY) {
+      Double.NEGATIVE_INFINITY
+    } else if (abs(a) == 0f) {
+      -Double.MIN_VALUE
+    } else {
+      cmath.nextafter(a, Double.NEGATIVE_INFINITY)
+    }
```

Comment:
This whole method can simply be `nextAfter(a, Double.NEGATIVE_INFINITY`. There's no need for the special cases for the specials.